### PR TITLE
Pull Request: chore/create-ghosts

### PR DESCRIPTION
**Pull request by stravid**
This is the first part of the ghost feature. It introduces the ghost on both client and server side. If the signed in user improves his highscore we create a new ghost and delete the old one.

I'm looking forward for suggestions on how the code can be improved!

### CODE REVIEW

File: app/assets/javascripts/slotcars/play/controllers/game_controller.js.coffee [hunk 1]
```diff
@@ -38,14 +38,31 @@ Play.GameController = Shared.BaseGameController.extend
...
+
+  loadHighscores: (callback) ->
+    @track.loadHighscores (highscores) =>
+      @onHighscoresLoaded highscores
+
+      callback() if callback?
+
+  saveGhost: ->
+    ghost = Shared.Ghost.createRecord
+      positions: @records
```
Comment (tombu) [inline app/assets/javascripts/slotcars/play/controllers/game_controller.js.coffee]: As I already mentioned some days ago the noun `record` is a little confusing here. What about something like `recordedPositions`?

File: app/assets/javascripts/slotcars/play/controllers/game_controller.js.coffee [hunk 2]
```diff
@@ -38,14 +38,31 @@ Play.GameController = Shared.BaseGameController.extend
...
+      track: @track
+      user: Shared.User.current
+      time: @get 'raceTime'
```
Comment (tombu) [inline app/assets/javascripts/slotcars/play/controllers/game_controller.js.coffee]: We shouldn´t need `get` here.. you neither use it in line 64 and in the `Recordable` mixin ..

File: spec/controllers/api/ghosts_controller_spec.rb [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+      post :create, :ghost => { :time => -42, :track_id => track.id, :positions => 'hellotom' }
+
+      response.should be_error
+    end
+
+    it 'should destroy the already present ghost' do
+      Ghost.create! :time => 40, :user_id => user.id, :track_id => track.id, :positions => 'yeahyeah'
+
+      sign_in user
+      new_time = 42
```
Comment (tombu) [inline spec/controllers/api/ghosts_controller_spec.rb]: It seems something is wrong here. It should be the other way round. Only save the ghost if the time is less (thous faster) ... :wink:

File: spec/javascripts/unit/slotcars/play/controllers/game_controller_spec.js.coffee [hunk 4]
```diff
@@ -253,3 +253,52 @@ describe 'Play.GameController (unit)', ->
...
+
+    afterEach ->
+      @highscoresMock.restore()
+
+      Shared.User.current = null
+
+    it 'should call saveGhost if the highscores got better', ->
+      sinon.stub @gameController, 'saveGhost'
+
+      @gameController.raceTime = 9
```
Comment (tombu) [inline spec/javascripts/unit/slotcars/play/controllers/game_controller_spec.js.coffee]: 9 ist faster than 10, so the ghost should be saved, shouldn´t it?

**Comment by tombu** [comment]
I like it much better now :wink: - :+1: for merge

**Comment by DominikGuzei** [comment]
we probably should merge the PR #179 first because it changes the way "isCarControlsEnabled" is handled

File: app/assets/javascripts/slotcars/shared/models/ghost.js.coffee [hunk 5]
```diff
@@ -0,0 +1,9 @@
+Shared.Ghost = DS.Model.extend
+
+  positions: DS.attr 'json'
+  time: DS.attr 'number'
+  user: DS.belongsTo 'Shared.User'
+  track: DS.belongsTo 'Shared.Track'
+
+  save: ->
+    Shared.ModelStore.commit()
```
Comment (DominikGuzei) [inline app/assets/javascripts/slotcars/shared/models/ghost.js.coffee]: one-liner?

File: spec/javascripts/unit/slotcars/play/controllers/game_controller_spec.js.coffee [hunk 6]
```diff
@@ -253,3 +253,56 @@ describe 'Play.GameController (unit)', ->
...
+
+      Shared.User.current = 
+        get: ->
+
+    afterEach ->
+      @highscoresMock.restore()
+
+      Shared.User.current = null
+
+    it 'should return if the highscore is better', ->
```
Comment (DominikGuzei) [inline spec/javascripts/unit/slotcars/play/controllers/game_controller_spec.js.coffee]: please split this spec into two (for true/false cases) -> also correct the name "should return TRUE if the highscore is better"

File: spec/javascripts/unit/slotcars/play/controllers/game_controller_spec.js.coffee [hunk 7]
```diff
@@ -253,3 +253,56 @@ describe 'Play.GameController (unit)', ->
...
+      @gameController.raceTime = 10
+      (expect @gameController.isNewHighscore()).toBe false
+
+      @gameController.raceTime = @highscoreTime
+      (expect @gameController.isNewHighscore()).toBe true
+
+  describe '#saveGhost', ->
+
+    beforeEach ->
+      sinon.spy Shared.Ghost, 'createRecord'
```
Comment (DominikGuzei) [inline spec/javascripts/unit/slotcars/play/controllers/game_controller_spec.js.coffee:285]: one-liner

File: spec/javascripts/unit/slotcars/play/controllers/game_controller_spec.js.coffee [hunk 8]
```diff
@@ -253,3 +253,56 @@ describe 'Play.GameController (unit)', ->
...
+
+    afterEach ->
+      Shared.Ghost.createRecord.restore()
```
Comment (DominikGuzei) [inline spec/javascripts/unit/slotcars/play/controllers/game_controller_spec.js.coffee:288]: one-liner

File: spec/javascripts/unit/slotcars/shared/lib/recordable_spec.js.coffee [hunk 9]
```diff
@@ -0,0 +1,49 @@
+
+describe 'Shared.Recordable', ->
+
+  beforeEach -> 
+    @recordable = Ember.Object.extend(Shared.Recordable).create()
+    @recordable._super = ->
+
+  describe '#update', ->
+
+    it 'should only record the car position if the controls are enabled', ->
```
Comment (DominikGuzei) [inline spec/javascripts/unit/slotcars/shared/lib/recordable_spec.js.coffee:12]: please split this spec into two (true/false)

**Comment by DominikGuzei** [comment]
apart from the minor issues it looks really sweet and awesome!! :+1: for merge!

**Pull request merged**